README.md
# Rancher Service Updater

[![CircleCI](https://img.shields.io/circleci/project/github/objectpartners/rancher-service-updater.svg?maxAge=0)](https://circleci.com/gh/objectpartners/rancher-service-updater/tree/master)
[![GitHub Release](https://img.shields.io/github/release/objectpartners/rancher-service-updater.svg?maxAge=0)](https://github.com/objectpartners/rancher-service-updater/releases)
[![Apache License 2.0](https://img.shields.io/badge/license-Apache_License_2.0-blue.svg)](https://github.com/objectpartners/rancher-service-updater/blob/master/LICENSE)

An Inversion of Control (IOC) service used to notify Rancher of new container 
images and execute updates to those services based on container labels.

## Configuring

* `AUTOUPDATE_ENABLE_LABEL` [`autoupdate.enable`] - Specifies the container label to query for automatic update enabling.
* `AUTOUPDATE_ENVIRONMENT_NAMES` [`[".*"]`] - An array  of regex patterns to match Rancher environment names against. Environment name must match a pattern for auto-updating to occur in that environment.
* `AUTOUPDATE_HTTP_PORT` [`8080`] - The port that the service updater listens on.
* `AUTOUPDATE_SLACK_WEBHOOK_URL` - The webhook URL to use for sending Slack notifications. If not specified, Slack messaging is disabled.
* `AUTOUPDATE_SLACK_BOT_NAME` - The bot name to send as for Slack messages.
* `CATTLE_ACCESS_KEY` - The API access key for Rancher.
* `CATTLE_SECRET_KEY` - The API secret key for Rancher.
* `CATTLE_URL` - The Rancher server URL.

## Enabling Auto-Updating for Rancher Environment

The Rancher Service Updater configures itself to only check certain environments for services to automatically update.
This is done in conjunction with the API restrictions based on the credentials passed to the container.
That is, if an environment specific key is used, then the updater will only ever try to update service within that environment.

If however, a global/user key is used, then the updater can update services in any environment that that key has access to.
By default the updater is configured to allow updating of services in any environment.
This is set via the `AUTOUDPATE_ENVIRONMENT_NAMES` environment variable which should contain a comma (`,`) separated list of regex patterns.
The default value is `.*`.

In conjunction with a global key, this property can be used to restrict the environments.
For example, assume a cluster with 3 environments: `dev`, `qa`, and `production`.
The updater is configured with a global key that has access to all 3 environments.
To restrict automatic updates to only the `dev` environment, set `AUTOUPDATE_ENVIRONMENT_NAMES=dev`.
To restrict automatic updates to both the `dev` and `qa` environments, set `AUTOUPDATE_ENVIRONMENT_NAMES=dev,qa`


## Enabling Auto-Updating for a Rancher Service

Configure a service to be automatically updated by adding a container label of `autoupdate.enabled=true`.
Alternatively, the label to check can be specified by setting the `AUTOUPDATE_ENABLE_LABEL` environment variable

### Determining if update is required

| Currently Deployed Version | Newly Published Version | Update? |
|:---------------------------|:------------------------|:--------|
| 1.0                        | 2.0                     | `true`  |
| 2.0                        | 1.3                     | `false` |
| 1.0                        | latest                  | `true`  |
| latest                     | latest                  | `true`  |
| latest                     | 1.0                     | `false` |

## Running Service Updater on Rancher

The Rancher Service Updater relies upon the standard environment variables for providing 
connection information to create a Rancher client instance.

If running this application as a container on a Rancher cluster itself, you can 
provide the following service labels to automatically provision the API 
credentials to the application.

* `io.rancher.container.create_agent=true`
* `io.rancher.container.agent.role=environment`

_NOTE:_ When using this method, the provisioned API credentials are tied to the
Rancher environment in which the service is deployed. Thus, it will only have access
to update services within that same environment.

_NOTE:_ If it is intended to auto-update services in multiple Rancher environments,
then the API configuration **must** be provided via environment variables.

## Triggering an upgrade.

Send a post to `/upgrade` with the following JSON payload:

```
{
  "docker_image": "docker:",
  "confirm": true,
  "start_first": false,
  "timeout": 30
}
```

* `docker_image` - the image path that is now available. Optionally start with `docker:`
* `confirm` - if the service upgrade should be confirmed/finished if successful
* `start_first` - Optional. Default of `false`. If true, then sets new services to be started before terminated old services.
* `timeout` - Optional. Timeout in seconds. Default of 30. Timeout for waiting for service upgrade to complete if `confirm = true`.

## Security

This service provides not mechanism for authentication/authorization. It is the responsibility of the user to properly secure 
this service such that unauthorized access is not available.


CHANGELOG.md
## 0.2 (Unreleased)

BACKWARDS INCOMPATIBILITIES / NOTES:

FEATURES:

IMPROVEMENTS

BUG FIXES:

## 0.1 (January 25, 2017)

BACKWARDS INCOMPATIBILITIES / NOTES:

* Initial Release

FEATURES:

* Initial Release

IMPROVEMENTS

* Initial Release

BUG FIXES:

* Initial Release


main_test.go
package main

import (
	"testing"

	"github.com/rancher/go-rancher/client"
)

func Test_basic(t *testing.T) {
	_ = &ServiceUpdater{
		Config: &Config{
			EnableLabel:      "autoupdate.enable",
			EnvironmentNames: []string{".*"},
			Port:             8080,
		},
		service: &mockService{},
		account: &mockAccount{},
	}

}

type mockService struct{}
type mockAccount struct{}

func (a *mockService) ById(id string) (*client.Service, error) {
	return nil, nil
}

func (a *mockService) List(opts *client.ListOpts) (*client.ServiceCollection, error) {
	return nil, nil
}

func (a *mockService) ActionFinishupgrade(service *client.Service) (*client.Service, error) {
	return nil, nil
}

func (a *mockService) ActionUpgrade(service *client.Service, serviceUpgrade *client.ServiceUpgrade) (*client.Service, error) {
	return nil, nil
}

func (a *mockAccount) List(opts *client.ListOpts) (*client.AccountCollection, error) {
	return nil, nil
}


main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/ahaynssen/slack-go-webhook"
	"github.com/objectpartners/rancher-service-updater/utils"
	"github.com/rancher/go-rancher/client"
)

type (
	//Config is the service configuration
	Config struct {
		EnableLabel      string
		EnvironmentNames []string
		Port             int
		CattleSecretKey  string
		CattleAccessKey  string
		CattleURL        string
		SlackWebhookURL  string
		SlackBotName     string
		Debug            bool
	}

	//ServiceUpdater is the service
	ServiceUpdater struct {
		Config *Config
		// client  *client.RancherClient
		service Service
		account Account
	}

	//UpdateCommand is payload for new image availability
	UpdateCommand struct {
		Image      string `json:"docker_image"`
		StartFirst bool   `json:"start_first"`
		Confirm    bool   `json:"confirm"`
		Timeout    int    `json:"timeout"`
	}

	//Service is Rancher Service interface
	Service interface {
		ById(id string) (*client.Service, error)
		List(opts *client.ListOpts) (*client.ServiceCollection, error)
		ActionFinishupgrade(*client.Service) (*client.Service, error)
		ActionUpgrade(*client.Service, *client.ServiceUpgrade) (*client.Service, error)
	}

	//Account is Rancher Environment interface
	Account interface {
		List(opts *client.ListOpts) (*client.AccountCollection, error)
	}
)

func main() {
	config := &Config{
		EnableLabel:      utils.GetEnvOrDefault("AUTOUPDATE_ENABLE_LABEL", "autoupdate.enable"),
		EnvironmentNames: utils.GetEnvOrDefaultArray("AUTOUPDATE_ENVIRONMENT_NAMES", []string{".*"}),
		Port:             utils.GetEnvOrDefaultInt("AUTOUPDATE_HTTP_PORT", 8080),
		CattleAccessKey:  os.Getenv("CATTLE_ACCESS_KEY"),
		CattleSecretKey:  os.Getenv("CATTLE_SECRET_KEY"),
		CattleURL:        os.Getenv("CATTLE_URL"),
		SlackWebhookURL:  os.Getenv("AUTOUPDATE_SLACK_WEBHOOK_URL"),
		SlackBotName:     utils.GetEnvOrDefault("AUTOUPDATE_SLACK_BOT_NAME", "rancher-service-updater"),
		Debug:            os.Getenv("DEBUG") != "",
	}
	serviceUpdater := &ServiceUpdater{
		Config: config,
	}
	serviceUpdater.init()
	serviceUpdater.listen()
}

func (s *ServiceUpdater) init() {
	c, err := client.NewRancherClient(&client.ClientOpts{
		AccessKey: s.Config.CattleAccessKey,
		SecretKey: s.Config.CattleSecretKey,
		Url:       s.Config.CattleURL,
	})
	if err != nil {
		log.Fatalf("Unable to create Rancher client: %s\n", err)
	}
	s.service = c.Service
	s.account = c.Account
}

func (s *ServiceUpdater) listen() {
	http.HandleFunc("/upgrade", s.upgrade)
	http.HandleFunc("/ping", s.ping)
	log.Printf("Started service on port %d\n", s.Config.Port)
	err := http.ListenAndServe(fmt.Sprintf(":%d", s.Config.Port), nil)
	if err != nil {
		log.Fatalf("Unable to start service on port %d\n", s.Config.Port)
	}
}

func (s *ServiceUpdater) ping(w http.ResponseWriter, r *http.Request) {
	w.Write([]byte("Pong!"))
	return
}

func (s *ServiceUpdater) upgrade(w http.ResponseWriter, r *http.Request) {
	command := UpdateCommand{
		StartFirst: false,
		Timeout:    30,
	}

	err := json.NewDecoder(r.Body).Decode(&command)
	if err != nil {
		log.Printf("%s\n", err.Error())
		utils.SendError(w, err.Error(), 400)
		return
	}
	txt, _ := json.Marshal(command)
	if s.Config.Debug {
		fmt.Printf("Received upgrade: %s", string(txt))
	}
	go s.upgradeService(command)
	w.WriteHeader(200)
	return
}

func (s *ServiceUpdater) upgradeService(command UpdateCommand) {
	if !strings.HasPrefix(command.Image, "docker:") {
		command.Image = fmt.Sprintf("docker:%s", command.Image)
	}
	idx := strings.LastIndex(command.Image, ":")
	wantedImage := command.Image[0 : idx-1]
	wantedVer := command.Image[idx:]

	services, err := s.service.List(&client.ListOpts{})
	if err != nil {
		fmt.Printf("Failed to list rancher services: %s\n", err)
		return
	}

	environments, err := s.account.List(&client.ListOpts{})
	if err != nil {
		fmt.Printf("Failed to get environments: %s\n", err)
		return
	}
	envs := make(map[string]string)
	for environments != nil {
		for _, env := range environments.Data {
			envs[env.Id] = env.Name
		}
		environments, err = environments.Next()
		if err != nil {
			fmt.Printf("Failed: %s\n", err)
			return
		}
	}

	var enabledLabel = s.Config.EnableLabel
	for services != nil {
		for _, svc := range services.Data {
			if s.Config.Debug {
				log.Printf("Checking service: %s", svc.Name)
			}
			if svc.LaunchConfig != nil {
				if enable, ok := svc.LaunchConfig.Labels[enabledLabel]; ok && enable != "false" {
					if s.Config.Debug {
						log.Printf("Attempting to update service %s\n", svc.Name)
					}
					idx := strings.LastIndex(svc.LaunchConfig.ImageUuid, ":")

					foundImage := svc.LaunchConfig.ImageUuid[0 : idx-1]
					foundVer := svc.LaunchConfig.ImageUuid[idx:]
					if utils.EnvironmentEnabled(envs[svc.AccountId], s.Config.EnvironmentNames) {
						if s.Config.Debug {
							log.Printf("Service %s Comparision: found-image %s, found-version %s, wanted-image %s, wanted-version %s\n", svc.Name, foundImage, foundVer, wantedImage, wantedVer)
						}
						if foundImage == wantedImage && ((foundVer < wantedVer) || (wantedVer == "latest")) {
							fmt.Println("Trying to upgrade...")
							err := s.doUpgrade(command, svc)
							if err != nil {
								fmt.Println(err.Error())
							} else {
								if command.Confirm {
									fmt.Println("Trying to confirm...")
									err := s.confirmUpgrade(command, svc)
									url := fmt.Sprintf("%s/env/%s/apps/stacks/%s", s.Config.CattleURL, svc.AccountId, svc.EnvironmentId)
									if err != nil {
										fmt.Printf("Unable to upgrade service %s: %s\n", svc.Name, err.Error())
										message := fmt.Sprintf("Unable to confirm upgrade to `%s`.\nCheck status at <%[2]s|%[1]s>", svc.Name, url)
										s.slackMessage("danger", message)
									} else {
										fmt.Printf("Upgraded %s to %s\n", svc.Name, command.Image)
										message := fmt.Sprintf("`%[1]s` has been successfully upgraded to `%[2]s` "+
											"in %[4]s\n View in Rancher here: <%[3]s|%[1]s>", svc.Name, wantedVer, url, envs[svc.AccountId])
										s.slackMessage("good", message)

									}
								}
							}
							continue
						} else if s.Config.Debug {
							log.Printf("Updating enabled from environment [%s], service [%s], but published version [%s] was not newer than current version [%s]\n", envs[svc.AccountId], svc.Name, wantedVer, foundVer)
						}
					} else if s.Config.Debug {
						log.Printf("Updating not enabled for environment %s\n", envs[svc.AccountId])
					}
				}
			}
		}
		services, _ = services.Next()
	}
}

func (s *ServiceUpdater) doUpgrade(command UpdateCommand, service client.Service) error {
	service.LaunchConfig.ImageUuid = command.Image
	upgrade := &client.ServiceUpgrade{}
	upgrade.InServiceStrategy = &client.InServiceUpgradeStrategy{
		LaunchConfig:           service.LaunchConfig,
		SecondaryLaunchConfigs: service.SecondaryLaunchConfigs,
		StartFirst:             command.StartFirst,
	}
	upgrade.ToServiceStrategy = &client.ToServiceUpgradeStrategy{}
	_, err := s.service.ActionUpgrade(&service, upgrade)
	return err
}

func (s *ServiceUpdater) confirmUpgrade(command UpdateCommand, service client.Service) error {
	srv, err := utils.Retry(func() (interface{}, error) {
		s, e := s.service.ById(service.Id)
		if e != nil {
			return nil, e
		}
		if s.State != "upgraded" {
			return nil, fmt.Errorf("Service not upgraded: %s\n", s.State)
		}
		return s, nil
	}, time.Duration(command.Timeout)*time.Second, 3*time.Second)
	if err != nil {
		return err
	}

	srv, err = s.service.ActionFinishupgrade(srv.(*client.Service))
	if err != nil {
		return err
	}
	fmt.Printf("Finished upgrade on %s\n", srv.(*client.Service).Name)
	return err
}

func (s *ServiceUpdater) slackMessage(status string, message string) {
	if s.Config.SlackWebhookURL != "" {
		attachment := slack.Attachment{Color: &status, Text: &message}
		mrkdwn := "text"
		attachment.AddMrkdwn(&mrkdwn)
		payload := slack.Payload{
			Username:    s.Config.SlackBotName,
			Attachments: []slack.Attachment{attachment},
		}
		printable, _ := json.Marshal(payload)
		fmt.Println(string(printable))
		err := slack.Send(s.Config.SlackWebhookURL, "", payload)
		if len(err) > 0 {
			fmt.Printf("error sending Slack message: %s\n", err)
		}
	}
}
